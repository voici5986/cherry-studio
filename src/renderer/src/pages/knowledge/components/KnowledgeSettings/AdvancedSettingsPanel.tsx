import InfoTooltip from '@renderer/components/InfoTooltip'
import { KnowledgeBase } from '@renderer/types'
import { Alert, InputNumber } from 'antd'
import { TriangleAlert } from 'lucide-react'
import { useTranslation } from 'react-i18next'

import { SettingsItem, SettingsPanel } from './styles'

interface AdvancedSettingsPanelProps {
  newBase: KnowledgeBase
  handlers: {
    handleChunkSizeChange: (value: number | null) => void
    handleChunkOverlapChange: (value: number | null) => void
    handleThresholdChange: (value: number | null) => void
  }
}

const AdvancedSettingsPanel: React.FC<AdvancedSettingsPanelProps> = ({ newBase, handlers }) => {
  const { t } = useTranslation()
  const { handleChunkSizeChange, handleChunkOverlapChange, handleThresholdChange } = handlers

  return (
    <SettingsPanel>
      <SettingsItem>
        <div className="settings-label">
          {t('knowledge.chunk_size')}
          <InfoTooltip title={t('knowledge.chunk_size_tooltip')} placement="right" />
        </div>
        <InputNumber
          style={{ width: '100%' }}
          min={100}
          value={newBase.chunkSize}
          placeholder={t('knowledge.chunk_size_placeholder')}
          onChange={handleChunkSizeChange}
          aria-label={t('knowledge.chunk_size')}
        />
      </SettingsItem>

      <SettingsItem>
        <div className="settings-label">
          {t('knowledge.chunk_overlap')}
          <InfoTooltip title={t('knowledge.chunk_overlap_tooltip')} placement="right" />
        </div>
        <InputNumber
          style={{ width: '100%' }}
          min={0}
          value={newBase.chunkOverlap}
          placeholder={t('knowledge.chunk_overlap_placeholder')}
          onChange={handleChunkOverlapChange}
          aria-label={t('knowledge.chunk_overlap')}
        />
      </SettingsItem>

      <SettingsItem>
        <div className="settings-label">
          {t('knowledge.threshold')}
          <InfoTooltip title={t('knowledge.threshold_tooltip')} placement="right" />
        </div>
        <InputNumber
          style={{ width: '100%' }}
          step={0.1}
          min={0}
          max={1}
          value={newBase.threshold}
          placeholder={t('knowledge.threshold_placeholder')}
          onChange={handleThresholdChange}
          aria-label={t('knowledge.threshold')}
        />
      </SettingsItem>

      <Alert
        message={t('knowledge.chunk_size_change_warning')}
        type="warning"
        showIcon
        icon={<TriangleAlert size={16} className="lucide-custom" />}
      />
    </SettingsPanel>
  )
}

export default AdvancedSettingsPanel
