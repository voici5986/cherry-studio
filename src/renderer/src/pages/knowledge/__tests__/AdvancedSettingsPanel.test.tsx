import type { KnowledgeBase, Model } from '@renderer/types'
import { fireEvent, render, screen } from '@testing-library/react'
import { beforeEach, describe, expect, it, vi } from 'vitest'

import AdvancedSettingsPanel from '../components/KnowledgeSettings/AdvancedSettingsPanel'

const mocks = vi.hoisted(() => {
  return {
    i18n: {
      t: (k: string) => {
        const translations: Record<string, string> = {
          'knowledge.chunk_size': '分块大小',
          'knowledge.chunk_overlap': '分块重叠',
          'knowledge.threshold': '检索相似度阈值',
          'knowledge.chunk_size_change_warning': '避免修改这个高级设置。'
        }
        return translations[k] || k
      }
    },
    handlers: {
      handleChunkSizeChange: vi.fn(),
      handleChunkOverlapChange: vi.fn(),
      handleThresholdChange: vi.fn()
    }
  }
})

vi.mock('@renderer/components/InfoTooltip', () => ({
  default: ({ title }: { title: string }) => <div>{mocks.i18n.t(title)}</div>
}))

vi.mock('react-i18next', () => ({
  useTranslation: () => ({
    t: mocks.i18n.t
  })
}))

vi.mock('lucide-react', () => ({
  TriangleAlert: () => <span>warning</span>
}))

vi.mock('antd', () => ({
  Alert: ({ message }: { message: string }) => <div role="alert">{message}</div>,
  InputNumber: ({ ref, value, onChange, placeholder, disabled, style, 'aria-label': ariaLabel }: any) => (
    <input
      ref={ref}
      type="number"
      data-testid="input-number"
      aria-label={ariaLabel}
      placeholder={placeholder}
      value={value ?? ''}
      onChange={(e) => onChange(e.target.valueAsNumber)}
      disabled={disabled}
      style={style}
    />
  )
}))

/**
 * 创建测试用的 KnowledgeBase 对象
 * @param overrides 可选的属性覆盖
 * @returns KnowledgeBase 对象
 */
function createKnowledgeBase(overrides: Partial<KnowledgeBase> = {}): KnowledgeBase {
  return {
    id: '1',
    name: 'Test KB',
    model: {
      id: 'test-model',
      provider: 'test-provider',
      name: 'Test Model',
      group: 'test'
    } as Model,
    items: [],
    created_at: Date.now(),
    updated_at: Date.now(),
    version: 1,
    chunkSize: 500,
    chunkOverlap: 200,
    threshold: 0.5,
    ...overrides
  }
}

describe('AdvancedSettingsPanel', () => {
  const mockBase = createKnowledgeBase()

  beforeEach(() => {
    vi.clearAllMocks()
  })

  describe('basic rendering', () => {
    it('should match snapshot', () => {
      const { container } = render(<AdvancedSettingsPanel newBase={mockBase} handlers={mocks.handlers} />)

      expect(container.firstChild).toMatchSnapshot()
    })
  })

  describe('handlers', () => {
    it('should call handlers when values are changed', () => {
      render(<AdvancedSettingsPanel newBase={mockBase} handlers={mocks.handlers} />)

      const chunkSizeInput = screen.getByLabelText('分块大小')
      fireEvent.change(chunkSizeInput, { target: { value: '600' } })
      expect(mocks.handlers.handleChunkSizeChange).toHaveBeenCalledWith(600)

      const chunkOverlapInput = screen.getByLabelText('分块重叠')
      fireEvent.change(chunkOverlapInput, { target: { value: '300' } })
      expect(mocks.handlers.handleChunkOverlapChange).toHaveBeenCalledWith(300)

      const thresholdInput = screen.getByLabelText('检索相似度阈值')
      fireEvent.change(thresholdInput, { target: { value: '0.6' } })
      expect(mocks.handlers.handleThresholdChange).toHaveBeenCalledWith(0.6)
    })
  })
})
