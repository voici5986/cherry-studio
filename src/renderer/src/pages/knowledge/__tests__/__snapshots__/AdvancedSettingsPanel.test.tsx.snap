// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`AdvancedSettingsPanel > basic rendering > should match snapshot 1`] = `
.c0 {
  padding: 0 16px;
}

.c1 {
  margin-bottom: 24px;
}

.c1 .settings-label {
  font-size: 14px;
  margin-bottom: 8px;
  display: flex;
  align-items: center;
  gap: 8px;
}

<div
  class="c0"
>
  <div
    class="c1"
  >
    <div
      class="settings-label"
    >
      分块大小
      <div>
        knowledge.chunk_size_tooltip
      </div>
    </div>
    <input
      aria-label="分块大小"
      data-testid="input-number"
      placeholder="knowledge.chunk_size_placeholder"
      style="width: 100%;"
      type="number"
      value="500"
    />
  </div>
  <div
    class="c1"
  >
    <div
      class="settings-label"
    >
      分块重叠
      <div>
        knowledge.chunk_overlap_tooltip
      </div>
    </div>
    <input
      aria-label="分块重叠"
      data-testid="input-number"
      placeholder="knowledge.chunk_overlap_placeholder"
      style="width: 100%;"
      type="number"
      value="200"
    />
  </div>
  <div
    class="c1"
  >
    <div
      class="settings-label"
    >
      检索相似度阈值
      <div>
        knowledge.threshold_tooltip
      </div>
    </div>
    <input
      aria-label="检索相似度阈值"
      data-testid="input-number"
      placeholder="knowledge.threshold_placeholder"
      style="width: 100%;"
      type="number"
      value="0.5"
    />
  </div>
  <div
    role="alert"
  >
    避免修改这个高级设置。
  </div>
</div>
`;
